generator/src/generator/namespace/test_framework.rs

```diff
@@ -0,0 +1,344 @@
+                    out,
+                    "{}: GenRandom::generate(rng),",
+                    to_rust_variable_name(field.name().unwrap())
+                );
+            }
+        });
+        outln!(out, "}}");
+    }
+
+    /// Emit the `GenRandom` implementation for a switch.
+    fn gen_random_switch(
+        &self,
+        out: &mut Output,
+        switch_case: &xcbdefs::SwitchField,
+        case_info: &[CaseInfo],
+    ) {
+        // for each case, generate a random value
+        match switch_case.kind {
+            xcbdefs::SwitchKind::BitCase => {
+                // it's going to be a list of cases, GenRandom is implemented for
+                // Option<T> so we should be fine here
+                outln!(out, "Self {{");
+                out.indented(|out| {
+                    for (c_info, case) in case_info.iter().zip(switch_case.cases.iter()) {
+                        with_field_name(c_info, case, |field_name| {
+                            outln!(
+                                out,
+                                "{}: GenRandom::generate(rng),",
+                                to_rust_variable_name(field_name)
+                            );
+                        });
+                    }
+                });
+                outln!(out, "}}");
+            }
+            xcbdefs::SwitchKind::Case => {
+                outln!(out, "match rng.usize(..{}) {{", case_info.len());
+                out.indented(|out| {
+                    for (i, (c_info, case)) in
+                        case_info.iter().zip(switch_case.cases.iter()).enumerate()
+                    {
+                        if i == case_info.len() - 1 {
+                            out!(out, "_ => ");
+                        } else {
+                            out!(out, "{} => ", i);
+                        }
+
+                        with_field_name(c_info, case, |field_name| {
+                            outln!(
+                                out,
+                                "Self::{}(GenRandom::generate(rng)),",
+                                to_rust_type_name(field_name)
+                            );
+                        });
+                    }
+                });
+                outln!(out, "}}");
+            }
+        }
+    }
+
+    /// Emit the declaration for a `fastrand::Rng`, in the `rng` slot.
+    fn emit_rng_declaration(&self, out: &mut Output, factor: u64) {
+        // we need to pick a seed to ensure deterministic tests
+        // in order to ensure a deterministic build, use a strategy
+        // derived from the name of the test framework
+        let mut seed = 1u64;
+        for b in self.name.bytes() {
+            seed = seed.overflowing_mul(b as u64).0;
+            if seed == 0 {
+                seed = 1;
+            }
+        }
+        seed = seed.saturating_add(factor);
+
+        // emit the declaration
+        outln!(out, "let rng = Rng::with_seed({});", seed);
```

Comment:
Why not just hardcode `42` as the seed?

Reply:
Because I'd like to test a wide array of structures, instead of a set of structures that, on a bit level, are very similar.

`fastrand` is programmed to emit numbers as `u64` that are later casted into other numbers (e.g. `i8`). So if you run `fastrand` to generate e.g. an event, which usually have very similar byte layouts with each other, every event tested will be about the same.

There are ways to circumvent this with higher quality RNG, but it's easier to just mix up the seed a bit.